packages/nimble/inst/CppCode/tensorflow.cpp

```diff
@@ -0,0 +1,145 @@
+        }                                  \
+    }
+
+#define NIM_ASSERT_REL(x, rel, y)                                              \
+    NIM_ASSERT3((x)rel(y), "Expected " #x " " #rel " " #y "; actual %d vs %d", \
+                (x), (y))
+#define NIM_ASSERT_EQ(x, y) NIM_ASSERT_REL(x, ==, y)
+#define NIM_ASSERT_LT(x, y) NIM_ASSERT_REL(x, <, y)
+#define NIM_ASSERT_LE(x, y) NIM_ASSERT_REL(x, <=, y)
+
+// This is passed as the deallocator of memory that is not owned.
+void fake_deallocator(void* data, size_t len, void* arg) {}  // Does nothing.
+
+NimTfRunner::NimTfRunner(const std::string& graphDefBase64,
+                         const std::vector<std::string>& inputNames,
+                         const std::vector<std::string>& outputNames)
+    : status_(TF_NewStatus()),
+      graph_(TF_NewGraph()),
+      session_(NULL),
+      inputs_(inputNames.size()),
+      input_values_(inputNames.size()),
+      outputs_(outputNames.size()),
+      output_values_(outputNames.size()),
+      input_pos_(0),
+      output_pos_(outputNames.size()) {
+    // Initialize graph.
+    {
+        std::string graphDef;
+        Base64::Decode(graphDefBase64, &graphDef);
+        TF_Buffer buffer;
+        buffer.data = graphDef.data();
+        buffer.length = graphDef.size();
+        TF_ImportGraphDefOptions* options = TF_NewImportGraphDefOptions();
+        TF_GraphImportGraphDef(graph_, &buffer, options, status_);
+        TF_CHECK_OK(status_);
+        TF_DeleteImportGraphDefOptions(options);
+    }
+
+    // Initialize session.
+    {
+        TF_SessionOptions* options = TF_NewSessionOptions();
+        session_ = TF_NewSession(graph_, options, status_);
+        TF_CHECK_OK(status_);
+        TF_DeleteSessionOptions(options);
+    }
+
+    // Set inputs
+    for (int i = 0; i < inputNames.size(); ++i) {
+        inputs_[i].oper =
+            TF_GraphOperationByName(graph_, inputNames[i].c_str());
+        inputs_[i].index = 0;
+    }
+
+    // Set outputs.
+    for (int i = 0; i < outputNames.size(); ++i) {
+        outputs_[i].oper =
+            TF_GraphOperationByName(graph_, outputNames[i].c_str());
+        outputs_[i].index = 0;
+    }
+}
+
+NimTfRunner::~NimTfRunner() {
+    NIM_ASSERT_EQ(input_pos_, 0);
+    NIM_ASSERT_EQ(output_pos_, outputs_.size());
+
+    // Clean up.
+    TF_CloseSession(session_, status_);
+    TF_CHECK_OK(status_);
+    TF_DeleteSession(session_, status_);
+    TF_CHECK_OK(status_);
+    TF_DeleteGraph(graph_);
+    TF_DeleteStatus(status_);
+}
+
+void NimTfRunner::setInput(NimArrBase<double>& nimArr) {
+    NIM_ASSERT1(!nimArr.isMap(), "Cannot handle mapped array");
+    NIM_ASSERT_LT(input_pos_, inputs_.size());
+    NIM_ASSERT_EQ(output_pos_, 0);
+    static std::vector<int64_t> dims;
+    const int n = nimArr.numDims();
+    dims.resize(n);
+    size_t byteSize = sizeof(double);
+    for (int i = 0; i < n; ++i) {
+        const int dim = nimArr.dimSize(i);
+        dims[n - i - 1] = dim;  // Note the transpose.
```

Comment:
`std::reverse_copy` would require iterator access to `NimArrBase::dimSize()`, which would be easiest if we reimplement `dimSize()` to use an array rather than a switch statement. But that seems well beyond the scope of this PR :wink: